CONTRIBUTING.md

```diff
@@ -0,0 +1,233 @@
+    metadata.Add("Exception", e.ToString());
+    context.Tracer.RelatedError(metadata, $"Failed to compare {hook.Name} version");
+  }
+  ```
+
+## Error Handling
+
+- *Fail fast: An error or exception that risks data loss or corruption should shut down VFS4G immediately*
+
+  Preventing data loss and repository corruption is critical.  If an error or exception occurs that could lead to data loss, it's better to shut down VFS4G than keep the repository mounted and risk corruption.
+
+- *Do not catch exceptions that are indicative of a programmer error (e.g. `ArgumentNullException`)*
+
+  Any exceptions that result from programmer error (e.g. `ArgumentNullException`) should be discovered as early in the development process as possible.  Avoid `catch` statements that would hide these errors (e.g. `catch(Exception)`).
+
+  The only exception to this rule is for [unhandled exceptions in background threads](#bgexceptions)
+
+- *Do not use exceptions for normal control flow*
+
+  Prefer to write code that avoids exceptions being thrown (e.g. the `TryXXX` pattern).  There are performance costs to using exceptions for control flow, and VFS4G code typically needs to handle errors where (and as soon as) they occur.
+   
+- *Provide the user with user-actionable messages whenever possible*
+
+  Don't tell a user what went wrong.  Help the user fix the problem.
+
+  Example:
+  > `"You can only specify --hydrate if the repository is mounted. Run 'gvfs mount' and try again."`
+
+## Background Threads
+
+- *Avoid using the thread pool (and avoid using async)*
+
+  `HttpRequestor.SendRequest` makes a [blocking call](https://github.com/Microsoft/VFSForGit/blob/4baa37df6bde2c9a9e1917fc7ce5debd653777c0/GVFS/GVFS.Common/Http/HttpRequestor.cs#L135) to `HttpClient.SendAsync`.  That blocking call consumes a thread from the managed thread pool.  Until that design changes, the rest of VFS4G must avoid using the thread pool unless absolutely necessary.  If the thread pool is required, any long running tasks should be moved to a separate thread managed by VFS4G itself (see [GitMaintenanceQueue](https://github.com/Microsoft/VFSForGit/blob/4baa37df6bde2c9a9e1917fc7ce5debd653777c0/GVFS/GVFS.Common/Maintenance/GitMaintenanceQueue.cs#L19) for an example).
+
+  Long-running or blocking work scheduled on the managed thread pool can prevent the normal operation of VFS4G.  For example, it could prevent downloading file sizes, loose objects, or file contents in a timely fashion.
+
+- <a id="bgexceptions"></a>*Catch all exceptions on long-running tasks and background threads*
+
+  Wrap all code that runs in the background thread in a top-level `try/catch(Exception)`.  Any exceptions caught by this handler should be logged, and then force VFS4G to terminate with `Environment.Exit`.  It's not safe to allow VFS4G to continue to run after an exception that would have been unhandled in a long-running task (created with `TaskCreationOption.LongRunning`) or in a background thread.
```

Comment:
Were there details / discussion to include regarding why `Environment.Exit` is preferred over allowing the unhandled exception to propagate?

Reply:
> Were there details / discussion to include regarding why Environment.Exit is preferred over allowing the unhandled exception to propagate?

I will add some details on this point to the doc.

We've been using `Environment.Exit` because we know that it will consistently bring down the process regardless of how the background thread was started (e.g. native thread, `new Thread(...)`, or `Task.Factory.StartNew(..., TaskCreationOptions.LongRunning)`.

What we had run into is that for some background thread types the unhandled exception would not bring down the process.  I suspect in the case of `TaskCreationOptions.LongRunning` the exception is being stored in the `Task` returned by `Task.Factory.StartNew`.

I think that once the investigation/changes for #757 are complete we can update this guideline as needed (e.g. we might land at something like "Propagate background thread exceptions and do not use `TaskCreationOptions.LongRunning` to create background threads").

I've left a note in #757 that the guidelines should be updated when that change is made.